ihmc-java-toolkit/src/main/java/us/ihmc/tools/thread/LoopingThread.java

```diff
@@ -0,0 +1,193 @@
+{
+   private final Runnable target;
+   private final Throttler throttler = new Throttler();
+   private volatile double loopPeriodLimit = -1.0;
+   private final AtomicReference<RunCommand> runCommand = new AtomicReference<>(RunCommand.INITIALIZE);
+
+   public LoopingThread() {
+      this(-1.0);
+   }
+
+   public LoopingThread(double loopFrequencyLimit)
+   {
+      super();
+      target = null;
+      limitLoopFrequency(loopFrequencyLimit);
+   }
+
+   public LoopingThread(double loopFrequencyLimit, String name) {
+      this(null, loopFrequencyLimit, name);
+   }
+
+   public LoopingThread(Runnable target, String name)
+   {
+      this(target, -1.0, name);
+   }
+
+   public LoopingThread(Runnable target, double loopFrequencyLimit, String name)
+   {
+      super(target, name);
+      this.target = target;
+      limitLoopFrequency(loopFrequencyLimit);
+   }
+
+   public LoopingThread(RunnableThatThrows targetThatThrows, ExceptionHandler exceptionHandler, String name)
+   {
+      this(targetThatThrows, exceptionHandler, -1.0, name);
+   }
+
+   public LoopingThread(RunnableThatThrows targetThatThrows, ExceptionHandler exceptionHandler, double loopFrequencyLimit, String name)
+   {
+      super(() -> ExceptionTools.handle(targetThatThrows, exceptionHandler), name);
+      target = () -> ExceptionTools.handle(targetThatThrows, exceptionHandler);
+      limitLoopFrequency(loopFrequencyLimit);
+   }
+
+   /**
+    * Limit the frequency of the execution loop.
+    * To un-limit the loop frequency, pass in a number less than or equal to 0.0.
+    * @param frequencyLimit The limit for the loop frequency.
+    *                       If zero or negative, the loop's frequency is not limited.
+    */
+   public void limitLoopFrequency(double frequencyLimit)
+   {
+      loopPeriodLimit = Conversions.hertzToSeconds(frequencyLimit);
+   }
+
+   /**
+    * Starts executing the {@link #runInLoop()} method in a loop.
+    */
+   @Override
+   public void start()
+   {
+      synchronized (runCommand)
+      {
+         initializeIfNeeded();
+
+         runCommand.set(RunCommand.LOOP);
+         runCommand.notify();
+      }
+   }
+
+   /**
+    * Signal the thread to execute the {@link #runInLoop()} method once, then pause.
+    */
+   public void runOnce()
+   {
+      synchronized (runCommand)
+      {
+         initializeIfNeeded();
+
+         runCommand.set(RunCommand.RUN_ONCE);
+         runCommand.notify();
+      }
+   }
+
+   private void initializeIfNeeded()
+   {
+      if (runCommand.compareAndSet(RunCommand.INITIALIZE, RunCommand.PAUSE))
+         super.start();
+   }
+
+   /**
+    * Pauses re-executing the {@link #runInLoop()} method upon its completion.
+    */
+   public void pause()
+   {
+      runCommand.set(RunCommand.PAUSE);
+   }
+
+   /**
+    * Finalizes the execution of {@link #runInLoop()} and stops the thread.
+    * The thread cannot be re-started after calling this method.
+    */
+   public void close()
```

Comment:
What about destroy which makes it clear in one word that it can be used anymore after calling.

Reply:
close() is convention in Java. I.e. Closable, AutoClosable

Reply:
Neither Thread, nor this class extends Closable. This is just using "close" as a word.  That word holds no meaning in this context, like it would for a file or stream.

Reply:
I would be okay with others, like dispose, endThread, etc

Reply:
I think within the context of Java `close()` also indicates that the object cannot be used anymore (like `Closable`, `AutoClosable`). I've been exploring using `close()` over `destroy()` to stick with Java convention. If you prefer `destroy()` I can change it to that

Reply:
What convention are you citing? The documentation of Closable and AutoClosable talk about releasing resources like files and socket handles. I think you're confusing it with dispose (Disposable) which is more for this purpose. 

Regardless, Java's Thread doesn't require any cleanup, it just ends. So the name of this method just needs to convey that the class can't be used anymore if this is called. What about exitForever()?

https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/Closeable.html
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/AutoCloseable.html
https://docs.oracle.com/middleware/12211/coherence/java-reference/com/oracle/common/base/Disposable.html

Reply:
Okay, so a Thread is a resource until it ends, right? So what about adding implements Disposable and using `dispose()` here.

Reply:
Sure, I can make it a `Disposable`

Reply:
Oh, we don't actually have the `Disposable` in the dependencies of `ihmc-java-toolkit`. 

I'm down to extending `AutoClosable` though. `Disposable` is just an `AutoClosable` which changes the name of the method to call when releasing resources.